5.0 updates to System.Text.Json docs (#21108)
dotnet/docs: docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

This article shows how to migrate from [Newtonsoft.Json](https://www.newtonsoft.com/json) to <xref:System.Text.Json>.

> The `System.Text.Json` namespace provides functionality for serializing to and deserializing from JavaScript Object Notation (JSON). The `System.Text.Json` library is included in the runtime for [.NET Core 3.0](https://aka.ms/netcore3download) and later versions. For other target frameworks, install the [System.Text.Json](https://www.nuget.org/packages/System.Text.Json) NuGet package. The package supports:

Author: In the discussions about what to call runtime-dependent vs. framework-dependent deployments, the term "shared framework" was called out specifically as a term we should avoid. 

cc @terrajobst

* Not supported, workaround is not practical or possible. If you rely on these `Newtonsoft.Json` features, migration will not be possible without significant changes.

::: zone pivot="dotnet-5-0"
| Newtonsoft.Json feature                               | System.Text.Json equivalent |
|-------------------------------------------------------|-----------------------------|
| Case-insensitive deserialization by default           | ✔️ [PropertyNameCaseInsensitive global setting](#case-insensitive-deserialization) |
| Camel-case property names                             | ✔️ [PropertyNamingPolicy global setting](system-text-json-how-to.md#use-camel-case-for-all-json-property-names) |
| Minimal character escaping                            | ✔️ [Strict character escaping, configurable](#minimal-character-escaping) |

> | `NullValueHandling.Ignore` global setting             | ✔️ [IgnoreNullValues global option](system-text-json-how-to.md#exclude-all-null-value-properties) |

Reviewer: For 5.0 we should use `DefaultIgnoreCondition`, i.e. `JsonIgnoreCondition.WhenWritingNull` instead.

| Allow comments                                        | ✔️ [ReadCommentHandling global setting](#comments) |
| Allow trailing commas                                 | ✔️ [AllowTrailingCommas global setting](#trailing-commas) |
| Custom converter registration                         | ✔️ [Order of precedence differs](#converter-registration-precedence) |
| No maximum depth by default                           | ✔️ [Default maximum depth 64, configurable](#maximum-depth) |
| `PreserveReferencesHandling` global setting           | ✔️ [ReferenceHandling global setting](#preserve-object-references-and-handle-loops) |
| `ReferenceLoopHandling` global setting                | ✔️ [ReferenceHandling global setting](#preserve-object-references-and-handle-loops) |

> | Serialize or deserialize numbers in quotes            | ✔️ [NumberHandling global setting](#quoted-numbers) |

Reviewer: There are also per-property/field and per-type settings.

Author: How do you allow quoted numbers on a per-property or per-type basis?

| Deserialize to immutable classes and structs          | ✔️ [JsonConstructor, C# 9 Records](#deserialize-to-immutable-classes-and-structs) |

> | Support for fields                                    | ✔️ [IncludeFields global setting](#public-and-non-public-fields) |

Reviewer: There's also a per-field inclusion setting.

| `NullValueHandling` setting on `[JsonProperty]`       | ✔️ [JsonIgnore attribute](#conditionally-ignore-a-property)  |
| `DefaultValueHandling` setting on `[JsonProperty]`    | ✔️ [JsonIgnore attribute](#conditionally-ignore-a-property)  |
| Deserialize `Dictionary` with non-string key          | ✔️ [Supported](#dictionary-with-non-string-key) |
| Support for non-public property setters and getters   | ✔️ [JsonInclude attribute](#internal-and-private-property-setters-and-getters) |
| Support for a broad range of types                    | ⚠️ [Some types require custom converters](#types-without-built-in-support) |
| Polymorphic serialization                             | ⚠️ [Not supported, workaround, sample](#polymorphic-serialization) |
| Polymorphic deserialization                           | ⚠️ [Not supported, workaround, sample](#polymorphic-deserialization) |
| Deserialize inferred type to `object` properties      | ⚠️ [Not supported, workaround, sample](#deserialization-of-object-properties) |
| Deserialize JSON `null` literal to non-nullable value types | ⚠️ [Not supported, workaround, sample](#deserialize-null-to-non-nullable-type) |

> | `[JsonConstructor]` attribute                         | ⚠️ [Not supported, workaround, sample](#specify-constructor-to-use) |

Reviewer: This is supported now.

During deserialization, `Newtonsoft.Json` does case-insensitive property name matching by default. The <xref:System.Text.Json> default is case-sensitive, which gives better performance since it's doing an exact match. For information about how to do case-insensitive matching, see [Case-insensitive property matching](system-text-json-how-to.md#case-insensitive-property-matching).

If you're using `System.Text.Json` indirectly by using ASP.NET Core, you don't need to do anything to get behavior like `Newtonsoft.Json`. ASP.NET Core specifies the settings for [camel-casing property names](system-text-json-how-to.md#use-camel-case-for-all-json-property-names) and case-insensitive matching when it uses `System.Text.Json`.

::: zone pivot="dotnet-5-0"

> ASP.NET Core also enables [quoted numbers](#quoted-numbers) by default.

Reviewer: ```suggestion
ASP.NET Core also enables deserializing [quoted numbers](#quoted-numbers) by default.
```

The JSON value could not be converted to System.String.
```

> ## Scenarios using JsonSerializer that are supported starting in 5.0

Author: Good point, I reorganized to put all JsonSerializer sections under one H2.

The JSON value could not be converted to System.String.
```

## Scenarios using JsonSerializer that are supported starting in 5.0

### Quoted numbers

::: zone pivot="dotnet-5-0"
`Newtonsoft.Json` can serialize or deserialize numbers represented by JSON strings (surrounded by quotes). For example, it can accept: `{"DegreesCelsius":"23"}` instead of `{"DegreesCelsius":23}`. To enable that behavior in <xref:System.Text.Json>, set <xref:System.Text.Json.JsonSerializerOptions.NumberHandling%2A?displayProperty=nameWithType> to <xref:System.Text.Json.Serialization.JsonNumberHandling.WriteAsString> or <xref:System.Text.Json.Serialization.JsonNumberHandling.AllowReadingFromString>.

> If you're using `System.Text.Json` indirectly by using ASP.NET Core, you don't need to do anything to get behavior like `Newtonsoft.Json`. ASP.NET Core specifies [web defaults](system-text-json-how-to.md#web-defaults-for-jsonserializeroptions) when it uses `System.Text.Json`, which allows quoted numbers.

Reviewer 2: Should it be "which allow" singular at the end?

Author: Could go either way -- "**specifies** web defaults, which allows..." or "specifies **web defaults**, which allow..."

<xref:System.Text.Json> supports only parameterless constructors. As a workaround, you can call whichever constructor you need in a custom converter. See the example for [Deserialize to immutable classes and structs](#deserialize-to-immutable-classes-and-structs).
::: zone-end

### Conditionally ignore a property

`Newtonsoft.Json` has several ways to conditionally ignore a property on serialization or deserialization:

* `DefaultContractResolver` lets you select properties to include or exclude, based on arbitrary criteria.
* The `NullValueHandling` and `DefaultValueHandling` settings on `JsonSerializerSettings` let you specify that all null-value or default-value properties should be ignored.
* The `NullValueHandling` and `DefaultValueHandling` settings on the `[JsonProperty]` attribute let you specify individual properties that should be ignored when set to null or the default value.

<xref:System.Text.Json> provides the following ways to omit properties while serializing:

::: zone pivot="dotnet-5-0"

* The [[JsonIgnore]](system-text-json-how-to.md#exclude-individual-properties) attribute on a property causes the property to be omitted from the JSON during serialization.

> * The [IgnoreNullValues](system-text-json-how-to.md#exclude-all-null-value-properties) global option lets you exclude all null-value properties.

Reviewer: ```suggestion
* The [DefaultIgnoreCondition](system-text-json-how-to.md#exclude-all-null-value-properties) global option lets you exclude all null-value properties.
```

Maybe we should merge this with the info on L273.

FYI we are marking `IgnoreNullValues` as obsolete in .NET 6.0.

::: zone-end

::: zone pivot="dotnet-core-3-1"
<xref:System.Text.Json> 3.1 only works with public properties. Custom converters can provide this functionality.
::: zone-end

### Preserve object references and handle loops

By default, `Newtonsoft.Json` serializes by value. For example, if an object contains two properties that contain a reference to the same `Person` object, the values of that `Person` object's properties are duplicated in the JSON.

`Newtonsoft.Json` has a `PreserveReferencesHandling` setting on `JsonSerializerSettings` that lets you serialize by reference:

* An identifier metadata is added to the JSON created for the first `Person` object.
* The JSON that is created for the second `Person` object contains a reference to that identifier instead of property values.

`Newtonsoft.Json` also has a `ReferenceLoopHandling` setting that lets you ignore circular references rather than throw an exception.

::: zone pivot="dotnet-5-0"

> To preserve references and handle circular references in <xref:System.Text.Json>, set <xref:System.Text.Json.JsonSerializerOptions.ReferenceHandler%2A?displayProperty=nameWithType> to <xref:System.Text.Json.Serialization.ReferenceHandler.Preserve%2A>. The `ReferenceHandler.Preserve` setting is equivalent to `ReferenceLoopHandling` = `true` and `PreserveReferencesHandling` = `true` in `Newtonsoft.Json`.

Reviewer 3: ```suggestion
To preserve references and handle circular references in <xref:System.Text.Json>, set <xref:System.Text.Json.JsonSerializerOptions.ReferenceHandler%2A?displayProperty=nameWithType> to <xref:System.Text.Json.Serialization.ReferenceHandler.Preserve%2A>. The `ReferenceHandler.Preserve` setting is equivalent to `PreserveReferencesHandling` = `PreserveReferencesHandling.All` in `Newtonsoft.Json`.
```

I wouldn't say that `ReferenceHandler.Preserve` is equivalent to Newtonsoft's `ReferenceLoopHandling.Ignore`, we also have plans to add Ignore as a new feature for 6.0.

Also, [`PreserveReferencesHandling`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_PreserveReferencesHandling.htm) is an enum, not a bool.

Some related `Newtonsoft.Json` features are not supported:

> * [ReferenceResolver](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonSerializer_ReferenceResolver.htm)

Reviewer: We expose our own `ReferenceResolver` type which allows folks to implement custom preserve handling logic.

* `NullValueHandling` is set to `Ignore`, and
* During deserialization, the JSON contains a null value for a non-nullable value type.

> In the same scenario, <xref:System.Text.Json> does throw an exception. (For the corresponding null-handling setting in `System.Text.Json`, see [Conditionally ignore a property](#conditionally-ignore-a-property).)

Reviewer: Should this link to notes about per-property ignoring or the global setting i.e. `DefaultIgnoreCondition.WhenWritingDefault`?

Author: I believe the equivalent here is JsonSerializerOptions.IgnoreNullValues, which I removed from the main how-to doc per your comment:

>I'm not sure we should advertise this. IgnoreNullValues is obsolete starting in .NET 6.0, and we couldn't think of when it makes sense to ignore a null value received when deserializing.

### Deserialize to immutable classes and structs

`Newtonsoft.Json` can deserialize to immutable classes and structs because it can use constructors that have parameters.

::: zone pivot="dotnet-5-0"

> In <xref:System.Text.Json>, use the x to specify use of a parameterized constructor. Records in C# 9 are also immutable and are supported as deserialization targets. For more information, see [Immutable types and Records](system-text-json-how-to.md#immutable-types-and-records).

Reviewer: > use the x

Hmm what did you mean to write here 😄

For more information about custom converters that recursively call `Serialize` or `Deserialize`, see the [Required properties](#required-properties) section earlier in this article.

> ### Internal and private property setters and getters

Reviewer: ```suggestion
### Non-`public` property setters and getters
```

here and elsewhere
